bitfield/impl/src/bitfield_specifier.rs

```diff
@@ -0,0 +1,120 @@
+    quote_spanned,
+};
+
+pub fn generate(input: TokenStream) -> TokenStream {
+    match generate2(input.into()) {
+        Ok(output) => output.into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+pub fn generate2(input: TokenStream2) -> syn::Result<TokenStream2> {
+    let input = syn::parse::<syn::DeriveInput>(input.into())?;
+    match input.data {
+        syn::Data::Enum(data_enum) => {
+            generate3(
+                syn::ItemEnum {
+                    attrs: input.attrs,
+                    vis: input.vis,
+                    enum_token: data_enum.enum_token,
+                    ident: input.ident,
+                    generics: input.generics,
+                    brace_token: data_enum.brace_token,
+                    variants: data_enum.variants,
+                }
+            )
+        }
+        syn::Data::Struct(_) => {
+            bail!(
+                input.ident,
+                "structs are not supported as bitfield specifiers",
+            )
+        }
+        syn::Data::Union(_) => {
+            bail!(
+                input.ident,
+                "unions are not supported as bitfield specifiers",
+            )
+        }
+    }
+}
+
+pub fn generate3(input: syn::ItemEnum) -> syn::Result<TokenStream2> {
+    let enum_ident = &input.ident;
+    let count_variants = input.variants.iter().count();
+    if !count_variants.is_power_of_two() {
+        bail!(
+            input,
+            "BitfieldSpecifier expected a number of variants which is a power of 2",
+        )
+    }
+    // We can take `trailing_zeros` returns type as the required amount of bits.
+    let bits = count_variants.trailing_zeros();
+    let bits_literal = syn::LitInt::new(bits as u64, syn::IntSuffix::None, Span2::call_site());
+
+    // let mut discriminator_mask = Punctuated::<syn::Expr, Token![&&]>::new();
+    let mut check_discriminants_tokens = quote! {};
+    for variant in input.variants.iter() {
+        match &variant.fields {
+            syn::Fields::Named(fields_named) => bail!(fields_named, "named fields in enum variants are not supported"),
+            syn::Fields::Unnamed(fields_unnamed) => bail!(fields_unnamed, "unnamed fields in enum variants are not supported"),
+            syn::Fields::Unit => {
+                let variant_ident = &variant.ident;
+                check_discriminants_tokens.extend(quote_spanned! { variant.span() =>
+                    enum #variant_ident {}
+                    impl checks::CheckDiscriminantInRange for #variant_ident {
+                        type CheckType = [(); ((#enum_ident::#variant_ident as usize) < (0x1 << #bits_literal)) as usize ];
+                    }
+                });
+            },
+        }
+    }
+
+    use crate::ident_ext::IdentExt as _;
+    let impl_mod_name = syn::Ident::from_str(&format!(
+        "impl_bitfield_checks_for_{}",
+        enum_ident.to_owned_string(),
+    ));
+
+    Ok(quote!{
+        mod #impl_mod_name {
```

Comment:
Best to avoid modules in macro-generated code. For example this causes problems here if someone writes:

```rust
fn f() {
    #[derive(BitfieldSpecifier)]
    enum E {
        ...
    }

    ...
}
```

where `super` would refer to the wrong thing -- the surrounding module, not the function's scope.

Reply:
I haven't thought about such a thing and used macros extensively in my other proc macros. This is really horrific! Thanks.

Reply:
Since I just found out that enums in Rust cannot have multiple variants associated to the same discriminant value I can potentially simplify this to using no module and no further types. Let's see.

Reply:
Also thank you a lot for your code review! :)

Reply:
Fixed this in the newest commit.